File: src/utils/bodyWeightAnalytics.js
const MS_PER_DAY = 24 * 60 * 60 * 1000;

export const BODY_WEIGHT_RANGES = ['7D', '30D', '3M', '1Y', 'All'];

const RANGE_DAYS = {
  '7D': 7,
  '30D': 30,
  '3M': 90,
  '1Y': 365,
};

function toDayKey(timestamp) {
  const d = new Date(timestamp);
  const year = d.getFullYear();
  const month = `${d.getMonth() + 1}`.padStart(2, '0');
  const day = `${d.getDate()}`.padStart(2, '0');
  return `${year}-${month}-${day}`;
}

function average(values) {
  if (!values.length) return null;
  const sum = values.reduce((total, value) => total + value, 0);
  return sum / values.length;
}

function startOfIsoWeek(timestamp) {
  const d = new Date(timestamp);
  const day = d.getDay();
  const diff = day === 0 ? -6 : 1 - day;
  d.setDate(d.getDate() + diff);
  d.setHours(0, 0, 0, 0);
  return d.getTime();
}

function startOfMonth(timestamp) {
  const d = new Date(timestamp);
  d.setDate(1);
  d.setHours(0, 0, 0, 0);
  return d.getTime();
}

function getAnchorTimestamp(entriesAsc) {
  if (!entriesAsc.length) return Date.now();
  return entriesAsc[entriesAsc.length - 1].timestamp;
}

export function normalizeBodyWeightEntries(rawEntries) {
  if (!Array.isArray(rawEntries)) return [];

  return rawEntries
    .map((entry, sourceIndex) => {
      const weight = Number.parseFloat(entry?.weight);
      const timestamp = new Date(entry?.date).getTime();
      if (!Number.isFinite(weight) || !Number.isFinite(timestamp)) return null;

      return {
        id: `${timestamp}-${sourceIndex}`,
        weight,
        timestamp,
        date: entry.date,
        sourceIndex,
      };
    })
    .filter(Boolean)
    .sort((a, b) => a.timestamp - b.timestamp);
}

export function filterEntriesByRange(entriesAsc, range) {
  if (!entriesAsc.length || range === 'All') return entriesAsc;
  const days = RANGE_DAYS[range];
  if (!days) return entriesAsc;

  const anchor = getAnchorTimestamp(entriesAsc);
  const cutoff = anchor - days * MS_PER_DAY;
  return entriesAsc.filter((entry) => entry.timestamp >= cutoff && entry.timestamp <= anchor);
}

export function dedupeLatestEntryByDay(entriesAsc) {
  const byDay = new Map();

  entriesAsc.forEach((entry) => {
    byDay.set(toDayKey(entry.timestamp), entry);
  });

  return [...byDay.values()].sort((a, b) => a.timestamp - b.timestamp);
}

export function downsampleSeries(points, maxPoints = 140) {
  if (points.length <= maxPoints) return points;
  const stride = (points.length - 1) / (maxPoints - 1);
  const result = [];
  for (let i = 0; i < maxPoints; i += 1) {
    const idx = Math.round(i * stride);
    result.push(points[idx]);
  }
  return result;
}

export function calculateBodyWeightSummary(allEntriesAsc, selectedEntriesAsc) {
  const latestOverall = allEntriesAsc[allEntriesAsc.length - 1] || null;
  const firstOverall = allEntriesAsc[0] || null;
  const latestSelected = selectedEntriesAsc[selectedEntriesAsc.length - 1] || latestOverall;
  const latestSelectedIndex = latestSelected
    ? allEntriesAsc.findIndex((entry) => entry.id === latestSelected.id)
    : -1;
  const previousEntry = latestSelectedIndex > 0 ? allEntriesAsc[latestSelectedIndex - 1] : null;
  const anchorTs = latestSelected?.timestamp || getAnchorTimestamp(allEntriesAsc);

  const getWindowAverage = (windowStart, windowEnd) => {
    const values = allEntriesAsc
      .filter((entry) => entry.timestamp > windowStart && entry.timestamp <= windowEnd)
      .map((entry) => entry.weight);
    return average(values);
  };

  const weekStart = startOfIsoWeek(anchorTs);
  const monthStart = startOfMonth(anchorTs);
  const weekEntries = allEntriesAsc.filter((entry) => entry.timestamp >= weekStart && entry.timestamp <= anchorTs);
  const monthEntries = allEntriesAsc.filter((entry) => entry.timestamp >= monthStart && entry.timestamp <= anchorTs);
  const recentAvg = getWindowAverage(anchorTs - 7 * MS_PER_DAY, anchorTs);
  const previousAvg = getWindowAverage(anchorTs - 14 * MS_PER_DAY, anchorTs - 7 * MS_PER_DAY);

  return {
    currentWeight: latestSelected?.weight ?? null,
    changeFromPrevious:
      latestSelected && previousEntry ? latestSelected.weight - previousEntry.weight : null,
    weeklyTrend:
      recentAvg != null && previousAvg != null ? recentAvg - previousAvg : null,
    weekChange:
      weekEntries.length >= 2
        ? weekEntries[weekEntries.length - 1].weight - weekEntries[0].weight
        : null,
    monthChange:
      monthEntries.length >= 2
        ? monthEntries[monthEntries.length - 1].weight - monthEntries[0].weight
        : null,
    totalChange:
      latestOverall && firstOverall ? latestOverall.weight - firstOverall.weight : null,
  };
}

File: app/src/main/java/com/ironlog/app/util/BodyWeightAnalytics.kt
package com.ironlog.app.util

import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import kotlin.math.ceil
import kotlin.math.roundToInt

private const val MS_PER_DAY: Long = 24L * 60L * 60L * 1000L
val BODY_WEIGHT_RANGES = listOf("7D", "30D", "3M", "1Y", "All")
private val RANGE_DAYS = mapOf("7D" to 7, "30D" to 30, "3M" to 90, "1Y" to 365)

data class RawBodyWeightEntry(val id: String? = null, val weight: Any? = null, val date: String? = null)
data class NormalizedBodyWeightEntry(
    val id: String,
    val weight: Double,
    val timestamp: Long,
    val date: String,
    val sourceIndex: Int,
)
data class BodyWeightSummary(
    val currentWeight: Double?,
    val changeFromPrevious: Double?,
    val weeklyTrend: Double?,
    val weekChange: Double?,
    val monthChange: Double?,
    val totalChange: Double?,
)

private fun toDayKey(timestamp: Long): String = Instant.ofEpochMilli(timestamp)
    .atZone(ZoneId.systemDefault()).toLocalDate().toString()

private fun average(values: List<Double>): Double? = if (values.isEmpty()) null else values.sum() / values.size

private fun startOfIsoWeek(timestamp: Long): Long {
    val zone = ZoneId.systemDefault()
    val d = Instant.ofEpochMilli(timestamp).atZone(zone).toLocalDate()
    val diff = if (d.dayOfWeek.value == 7) -6 else 1 - d.dayOfWeek.value
    return d.plusDays(diff.toLong()).atStartOfDay(zone).toInstant().toEpochMilli()
}

private fun startOfMonth(timestamp: Long): Long {
    val zone = ZoneId.systemDefault()
    val d = Instant.ofEpochMilli(timestamp).atZone(zone).toLocalDate().withDayOfMonth(1)
    return d.atStartOfDay(zone).toInstant().toEpochMilli()
}

private fun getAnchorTimestamp(entriesAsc: List<NormalizedBodyWeightEntry>): Long =
    entriesAsc.lastOrNull()?.timestamp ?: System.currentTimeMillis()

fun normalizeBodyWeightEntries(rawEntries: List<RawBodyWeightEntry>?): List<NormalizedBodyWeightEntry> {
    if (rawEntries == null) return emptyList()
    return rawEntries.mapIndexedNotNull { sourceIndex, entry ->
        val weight = when (val w = entry.weight) {
            is Number -> w.toDouble()
            is String -> w.toDoubleOrNull()
            else -> null
        }
        val timestamp = entry.date?.let { runCatching { Instant.parse(it).toEpochMilli() }.getOrNull() }
        if (weight == null || timestamp == null || !weight.isFinite()) null
        else NormalizedBodyWeightEntry(
            id = entry.id ?: "$timestamp-$sourceIndex",
            weight = weight,
            timestamp = timestamp,
            date = entry.date,
            sourceIndex = sourceIndex,
        )
    }.sortedBy { it.timestamp }
}

fun filterEntriesByRange(entriesAsc: List<NormalizedBodyWeightEntry>, range: String): List<NormalizedBodyWeightEntry> {
    if (entriesAsc.isEmpty() || range == "All") return entriesAsc
    val days = RANGE_DAYS[range] ?: return entriesAsc
    // Anchored to now, matching the JS fix.
    val cutoff = System.currentTimeMillis() - days * MS_PER_DAY
    return entriesAsc.filter { it.timestamp >= cutoff }
}

fun dedupeLatestEntryByDay(entriesAsc: List<NormalizedBodyWeightEntry>): List<NormalizedBodyWeightEntry> {
    val byDay = LinkedHashMap<String, NormalizedBodyWeightEntry>()
    entriesAsc.forEach { byDay[toDayKey(it.timestamp)] = it }
    return byDay.values.sortedBy { it.timestamp }
}

fun <T> downsampleSeries(points: List<T>, maxPoints: Int = 140): List<T> {
    if (points.size <= maxPoints) return points
    val stride = (points.size - 1).toDouble() / (maxPoints - 1).toDouble()
    return List(maxPoints) { i -> points[(i * stride).roundToInt().coerceIn(points.indices)] }
}

fun calculateBodyWeightSummary(
    allEntriesAsc: List<NormalizedBodyWeightEntry>,
    selectedEntriesAsc: List<NormalizedBodyWeightEntry>,
): BodyWeightSummary {
    val latestOverall = allEntriesAsc.lastOrNull()
    val firstOverall = allEntriesAsc.firstOrNull()
    val latestSelected = selectedEntriesAsc.lastOrNull() ?: latestOverall
    val latestSelectedIndex = latestSelected?.let { ls -> allEntriesAsc.indexOfFirst { it.id == ls.id } } ?: -1
    val previousEntry = if (latestSelectedIndex > 0) allEntriesAsc[latestSelectedIndex - 1] else null
    val anchorTs = latestSelected?.timestamp ?: getAnchorTimestamp(allEntriesAsc)

    fun getWindowAverage(windowStart: Long, windowEnd: Long): Double? {
        return average(allEntriesAsc.filter { it.timestamp > windowStart && it.timestamp <= windowEnd }.map { it.weight })
    }

    val weekStart = startOfIsoWeek(anchorTs)
    val monthStart = startOfMonth(anchorTs)
    val weekEntries = allEntriesAsc.filter { it.timestamp >= weekStart && it.timestamp <= anchorTs }
    val monthEntries = allEntriesAsc.filter { it.timestamp >= monthStart && it.timestamp <= anchorTs }
    val recentAvg = getWindowAverage(anchorTs - 7 * MS_PER_DAY, anchorTs)
    val previousAvg = getWindowAverage(anchorTs - 14 * MS_PER_DAY, anchorTs - 7 * MS_PER_DAY)

    return BodyWeightSummary(
        currentWeight = latestSelected?.weight,
        changeFromPrevious = if (latestSelected != null && previousEntry != null) latestSelected.weight - previousEntry.weight else null,
        weeklyTrend = if (recentAvg != null && previousAvg != null) recentAvg - previousAvg else null,
        weekChange = if (weekEntries.size >= 2) weekEntries.last().weight - weekEntries.first().weight else null,
        monthChange = if (monthEntries.size >= 2) monthEntries.last().weight - monthEntries.first().weight else null,
        totalChange = if (latestOverall != null && firstOverall != null) latestOverall.weight - firstOverall.weight else null,
    )
}
